lib/url_to_subscriber_list_criteria.rb

```diff
@@ -0,0 +1,157 @@
+                 { "links" => from_params.merge("people" => [path_match[1]]) } # this is not yet implemented
+               elsif (path_match = @url.path.match(%r{^/government/ministers/(.*)\.atom$}))
+                 { "links" => from_params.merge("roles" => [path_match[1]]) } # this is not yet implemented
+               elsif (path_match = @url.path.match(%r{^/government/organisations/(.*)\.atom$}))
+                 { "links" => from_params.merge("organisations" => [path_match[1]]) }
+               elsif (path_match = @url.path.match(%r{^/government/topical-events/(.*)\.atom$}))
+                 { "links" => from_params.merge("topical_events" => [path_match[1]]) }
+               elsif (path_match = @url.path.match(%r{^/government/topics/(.*)\.atom$}))
+                 { "links" => from_params.merge(topic_map([path_match[1]]) => [path_match[1]]) }
+               elsif @url.path =~ %r{/government/feed}
+                 { 'links' => from_params }
+
+               else
+                 raise UnprocessableUrl, @url.to_s
+               end
+
+      if result.fetch("links", {})["publication_filter_option"]
+        result[GOVERNMENT_SUPERTYPE] = result["links"].delete("publication_filter_option")
+      end
+      if result.fetch("links", {})["announcement_filter_option"]
+        result[GOVERNMENT_SUPERTYPE] = result["links"].delete("announcement_filter_option")
+      end
+
+      result
+    end
+  end
+
+  def from_params
+    return {} if @url.query.blank?
+
+    result = Rack::Utils.parse_nested_query(@url.query)
+    {
+      'departments' => 'organisations',
+      'topics' => method(:topic_map),
+    }.each do |from_key, to_key|
+      next unless result.key?(from_key)
+
+      if to_key.is_a?(String)
+        result[to_key] = result.delete(from_key)
+      else
+        values = result.delete(from_key)
+        result[to_key.call(values)] = values
+      end
+    end
+    result
+  end
+
+  def topic_map(values)
+    @static_data.topical_event?(values) ? 'topical_events' : 'policy_areas'
+  end
+
+  def lookup_content_id(key, slug)
+    @static_data.content_id(key, slug).tap do |value|
+      @missing_lookup = "#{key}: #{slug}" if value == MISSING_LOOKUP
+    end
+  end
+
+  # Static data can currently be injected into the mapping process in two forms.
+  # This was done to allow optimisation of the process depending on the use case.
+  # - BulkStaticData which is desigend to be used with the mass migration of data
+  #   and can be deleted once the migration is complete.
+  # - StaticData which uses standard DB queries and is expected to be used when
+  #   converting individual URLs.
+  class BulkStaticData
+    def topical_event?(values)
+      @topical_events ||= TopicalEvent.pluck(:slug)
+      (@topical_events & values).any?
+    end
+
+    def content_id(key, slug)
+      case key
+      when "world_locations"
+        @world_location_lookups ||= Hash[WorldLocation.pluck(:slug, :content_id)]
```

Comment:
Sorry, I wasn't clear here before - when I said plural I meant the things, not `lookups` - so `world_locations_lookup` etc 🗺 👀 ⬆️